### PR TITLE
[quant][graphmode][fx] Add default_reuse_input_qconfig for ops like reshape

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69249

Summary:
This PR added default_reuse_input_qconfig and default_reuse_input_observer which is used
when we want to configure an operator to reuse the observer from input, if the input
Tensor for the operator is not observed, we will not observe the output of this operator either,
if the input Tensor is observed, we will observe the output of the operator with the same observer.

e.g.

```
x1 = x0.reshape()
```
if reshape is configured with default_reuse_input_qconfig:
1. if x0 is observed with observer_0, we'll observe x1 with the same observer instance
2. if x0 is not observed, we won't observe x1 either

Test Plan:
```
python test/test_quantization.py TestQuantizeFx.test_reuse_input_qconfig
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32774723](https://our.internmc.facebook.com/intern/diff/D32774723)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).

File: torch/ao/quantization/fx/qconfig_utils.py [hunk 1]
```diff
@@ -237,9 +237,12 @@ def generate_qconfig_map(
 
         elif node.op == "call_method":
             module_path, module_type = node_name_to_scope[node.name]
+            # first use node.target (string) to get the qconfig
+            qconfig = maybe_adjust_qconfig_for_module_type_or_name(
+                qconfig_dict, node.target, module_path, global_qconfig)
```
Comment (vkuzo) [inline torch/ao/quantization/fx/qconfig_utils.py:242]: is this related to this PR?  If yes, any chance we can make the comment more descriptive?

File: torch/ao/quantization/qconfig.py [hunk 2]
```diff
@@ -170,6 +172,12 @@ def __new__(cls, activation=torch.nn.Identity, weight=torch.nn.Identity):
 Fused version of `default_qat_config`, has performance benefits.
 """
 
+default_reuse_input_qconfig = QConfig(activation=default_reuse_input_observer,
+                                      weight=default_reuse_input_observer)
```
Comment (vkuzo) [inline torch/ao/quantization/qconfig.py:176]: would it make sense to set weight to `None`?  Not sure if there are cases where the right thing to do for current op is to reuse the weight observer of the previous op.

**Review by vkuzo — APPROVED** [review]
thanks for working on this!

File: torch/ao/quantization/fx/qconfig_utils.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/qconfig_utils.py:242]: yeah this is to enable that we can configure "object_type" with call_method strings, this functionality is missing previously and looks like no one has used this before

File: torch/ao/quantization/qconfig.py (hunk 2 above)
Comment (jerryzh168) [reply torch/ao/quantization/qconfig.py:176]: we don't have None observers right now I think, maybe I can set it to a NoopObserver

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32774723).
